### ISSUE TITLE
[Bug] System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.XXXRenderer from native handle

### ISSUE THREAD

**Issue by lassana** [xamarin/Xamarin.Forms#5894]
### Description

The title speaks for itself.

We observe this crash in *all* our apps. Sometimes it's `ScrollViewRenderer`, sometimes `MasterDetailPageRenderer`. In this particular sample it's usually `PageRenderer` but sometimes `LabelRenderer`.

Any suggestion how to avoid this crash would be highly appreciated.

### Steps to Reproduce

1. Open attached project, build & and run in Release configuration
2. Press the physical "Back" button so the activity is closed and the home screen appears (but the app is not closed!)
3. Launch the app again by tapping on app icon on the home screen.

### Expected Behavior

The main page is successfully re-created and the app works.

### Actual Behavior

The app crashes. Stacktrace:

```
android.runtime.JavaProxyThrowable: System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.PageRenderer from native handle 0xbfb0ff0c (key_handle 0xfa1a17a). ---> System.MissingMethodException: No constructor found for Xamarin.Forms.Platform.Android.PageRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership) ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
   --- End of inner exception stack trace ---
  at Java.Interop.TypeManager.CreateProxy (System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00055] in <1bab76874570475da27595819aee3371>:0 
  at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00116] in <1bab76874570475da27595819aee3371>:0 
   --- End of inner exception stack trace ---
  at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00182] in <1bab76874570475da27595819aee3371>:0 
  at Java.Lang.Object.GetObject (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type type) [0x000c1] in <1bab76874570475da27595819aee3371>:0 
  at Java.Lang.Object._GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00017] in <1bab76874570475da27595819aee3371>:0 
  at Java.Lang.Object.GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00000] in <1bab76874570475da27595819aee3371>:0 
  at Java.Lang.Object.GetObject[T] (System.IntPtr jnienv, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00006] in <1bab76874570475da27595819aee3371>:0 
  at Android.Views.View.n_OnDetachedFromWindow (System.IntPtr jnienv, System.IntPtr native__this) [0x00000] in <1bab76874570475da27595819aee3371>:0 
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.40(intptr,intptr)
--- End of stack trace from previous location where exception was thrown ---

  at Java.Interop.JniEnvironment+InstanceMethods.CallBooleanMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00069] in <d7f2b9bd15c64bcb95193fbdfb36a862>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeAbstractBooleanMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00014] in <d7f2b9bd15c64bcb95193fbdfb36a862>:0 
  at Android.Support.V4.App.FragmentManagerInvoker.ExecutePendingTransactions () [0x0000a] in <8f9b95e50b74401b836c79405557e90a>:0 
  at Xamarin.Forms.Platform.Android.FragmentManagerExtensions.ExecutePendingTransactionsEx (Android.Support.V4.App.FragmentManager fragmentTransaction) [0x00000] in <e11d74a01fc148949d5e89708470f54c>:0 
  at Xamarin.Forms.Platform.Android.AppCompat.NavigationPageRenderer.Dispose (System.Boolean disposing) [0x00290] in <e11d74a01fc148949d5e89708470f54c>:0 
  at Java.Lang.Object.Dispose () [0x00000] in <1bab76874570475da27595819aee3371>:0 
  at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.LoadApplication (Xamarin.Forms.Application application) [0x0026c] in <e11d74a01fc148949d5e89708470f54c>:0 
  at MvvmCross.Forms.Platforms.Android.Views.MvxFormsAppCompatActivity.InitializeApplication () [0x00007] in <4ea0c9ab37e14fd8aa188ddf8fa8736e>:0 
  at MvvmCross.Forms.Platforms.Android.Views.MvxFormsAppCompatActivity.RunAppStart (Android.OS.Bundle bundle) [0x0002a] in <4ea0c9ab37e14fd8aa188ddf8fa8736e>:0 
  at MvvmCross.Forms.Platforms.Android.Views.MvxFormsAppCompatActivity.OnCreate (Android.OS.Bundle bundle) [0x00028] in <4ea0c9ab37e14fd8aa188ddf8fa8736e>:0 
  at MyXamarinFormsApp.Droid.MainActivity.OnCreate (Android.OS.Bundle bundle) [0x00014] in <479aa9a0f42f4fe4af3e53fc8df13d8c>:0 
  at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_ (System.IntPtr jnienv, System.IntPtr native
```

### Basic Information

- Version with issue: Xamarin.Forms 3.6.0.293080
- Last known good version: I don't know. _I feel_ like 3.5 was more stable.
- IDE: VS 4 Mac 8.0
- Platform Target Frameworks: 
  - Android: 9.2.0.5 (Visual Studio Community)
- Nuget Packages: Xamarin.Forms 3.6.0.293080, MvvmCross 6.2.3.
- Affected Devices: All Android devices. Confirmed on Xiaomi RedMi 3S with Android 6.0. 

### Reproduction Link

[MyXamarinFormsApp.zip](https://github.com/xamarin/Xamarin.Forms/files/3070435/MyXamarinFormsApp.zip)

**Comment by samhouts** [comment]
@lassana I tried the sample app on my OnePlus 3T. I had to enabled x64 to deploy it. I don't get the crash you described, but I wonder if that's a consequence of the ABI. Are you able to confirm that it still crashes if you enable x64? Thanks!

**Comment by lassana** [comment]
@samhouts I'm not quite sure what you enabled because all ABIs are already enabled in the project: 
![Screen Shot 2019-04-12 at 9 41 52 AM](https://user-images.githubusercontent.com/2004960/56017207-395c1100-5d07-11e9-873c-0a9afa15ed47.png)

```
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    ...
    ....
    <AndroidSupportedAbis>armeabi-v7a;x86;arm64-v8a;x86_64</AndroidSupportedAbis>
  </PropertyGroup>
```

I also ran this on the `x86_64` emulator and the crash also occurs there, here is a video:

![ScreenRecording mov](https://user-images.githubusercontent.com/2004960/56017715-ca7fb780-5d08-11e9-8601-9fdf20c613c2.gif)

**Comment by samhouts** [comment]
Ah, I see! I ran it in debug mode and did not experience the crash. If this is happening in Release mode, it is likely a problem with the Linker removing one of your classes. I'll take a look soon and let you know if I find anything. Thanks!

**Comment by kvpt** [comment]
I don't know how MvvmCross work but I did have exactly the same issue with the same reproduction in my app because I was using a static App object.
When you click on the back arrow, the activity is destroyed and when you click after that on your app icon it's restart the app with a new activity.
If you use a static app object, the callback of main page change is called twice, one by activity and this cause the crash.

I made the PR #4707 to fix this.
Can you try if this patched version of Xamarin.Forms fix your issue ?
[xamarin.forms.3.6.3.zip](https://github.com/xamarin/Xamarin.Forms/files/3045663/xamarin.forms.3.6.3.zip)

Change the extension from .zip .to nupkg (because github limit extension type allowed).
To install the local package you can use the differents methods listed [here](https://stackoverflow.com/a/14601311).

It can also be another issue, because I have remaining crashs of this type but a lot less than before.

**Comment by lassana** [comment]
@kvpt I tried your package, unfortunately it doesn't work well. The app no longer crashes however the layout cannot be displayed for some reason.

![ScreenRecording mov](https://user-images.githubusercontent.com/2004960/56118117-4c264e00-5f72-11e9-9870-1a1aeab7f1af.gif)

[MyXamarinFormsApp-XF-3.6.3.zip](https://github.com/xamarin/Xamarin.Forms/files/3079161/MyXamarinFormsApp-XF-3.6.3.zip)

**Comment by kvpt** [comment]
@lassana Thanks for testing it.
I will try your sample and see if I find the issue with the fix.

**Comment by pablogupi** [comment]
Hello!
In my case, error occurs when receive Push Notification . If I click in notification and app stay in background, crash immediately and appear this error:

System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.PageRenderer from native handle 0xbf8e022c (key_handle 0xbc91b18). ---> System.MissingMethodException: No constructor found for Xamarin.Forms.Platform.Android.PageRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership) ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.

**Comment by thomashoef** [comment]
Also experiencing this issue in debug mode. The app launches ok - then go to droid settings, change the font size, switch back to the Xam Forms app. 
The app then (in launchmode singletop or singleinstance, does not matter) crashes:

05-15 21:04:22.615 I/MonoDroid(24953): UNHANDLED EXCEPTION:
05-15 21:04:22.630 I/MonoDroid(24953): System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.PageRenderer from native handle 0x65 (key_handle 0x70d5535). ---> System.MissingMethodException: No constructor found for Xamarin.Forms.Platform.Android.PageRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership) ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
05-15 21:04:22.630 I/MonoDroid(24953):    --- End of inner exception stack trace ---
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Interop.TypeManager.CreateProxy (System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00055] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00116] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):    --- End of inner exception stack trace ---
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00182] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Lang.Object.GetObject (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type type) [0x000c1] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Lang.Object._GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00017] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Lang.Object.GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00000] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Android.Views.ViewGroup.GetChildAt (System.Int32 index) [0x00036] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Xamarin.Forms.Platform.Android.AppCompat.Platform.SetPageInternal (Xamarin.Forms.Page newRoot) [0x0001a] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:301 
05-15 21:04:22.630 I/MonoDroid(24953):   at Xamarin.Forms.Platform.Android.AppCompat.Platform+<>c__DisplayClass36_0.<SetPage>b__0 () [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:269 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00009] in <44fafc2e0f1847ee827456a85a04372d>:0 
05-15 21:04:22.630 I/MonoDroid(24953):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.28(intptr,intptr)

**Comment by kvpt** [comment]
@lassana sorry for the delay.
I tried the version you posted with an emulator on marshmallow and I don't reproduce the blank screen.
![nocrash](https://user-images.githubusercontent.com/1446221/58999417-049ba000-8806-11e9-8c7c-b636386438db.gif)
I also forced the activity destruction in the developer settings to be sure but this seems to work as expected.

I know that certains users of my app reported that issue too but I never succeed to reproduce it.

I'm on windows your on mac, but I don't think this make a difference for this issue.

What is the configuration or your emulator ?
![image](https://user-images.githubusercontent.com/1446221/58999568-8f7c9a80-8806-11e9-8f4e-c883f27ba26d.png)

@pablogupi Have you a repro to test ?

@thomashoef From your stacktrace it is the issue that I fixed (method SetPageInternal)
Can you test with the preview version of Xamarin.Forms 4.1 on nuget ? The fix is included in it.

**Comment by lassana** [comment]
@kvpt My configuration is basically the same as yours. FYI: I tried `4.1.0.483098-pre1` and got blank screen.
[MyXamarinFormsApp.zip](https://github.com/xamarin/Xamarin.Forms/files/3259764/MyXamarinFormsApp.zip)

<details>
  <summary>Screencast</summary>
  
![ScreenRecording2019-06-06at5 54 22AM mov](https://user-images.githubusercontent.com/2004960/59003939-8bf60d00-8820-11e9-83a2-639e51a959e7.gif)

  
</details>

**Comment by MitchBomcanhao** [comment]
I have been getting this issue (at least) on versions 4.0.0 and 4.0.0 sr2, with the LabelRenderer being the thing that pops up as the missing method.

by process of elimination, the property in the labels that was triggering the crash was the vertical text alignment.


```
<Label Grid.Row="1"
FontAttributes="Bold"
FontSize="{x:Static local:AppConstants.MediumSize}"                                                  
LineBreakMode="TailTruncation"
MaxLines="2"
Text="{Binding Label, Converter={StaticResource TextCaseConverter}}"
VerticalOptions="End"
VerticalTextAlignment="End" />
```
Removing the text alignment prevents the crash.

for what its worth, this label lives in an ItemTemplate with a structure similar to:
```
<DataTemplate>
    <Grid>
        <Button />
        <Grid>
            <Grid.RowDefinitions>
                <RowDefinition Height="Auto" />
                <RowDefinition Height="Auto" />
            </Grid.RowDefinitions>
            <Label Grid.Row=1/>
        </Grid>
    </Grid>
</DataTemplate>
```

**Comment by kvpt** [comment]
I'm sure it's caused by the dispose process.
I have made a new patched 3.6 version which contains more fix related to dispose.
The nuget is in the zip => [xamarin.forms.3.6.6.zip](https://github.com/xamarin/Xamarin.Forms/files/3277034/xamarin.forms.3.6.6.zip)

@lassana If you want to test it (delete bin/obj first be sure)

@MitchBomcanhao Unfortunately I have not done a release for the branch v4 yet for you to test.
I think it's logic that the vertical alignment trigger the crash because it set the Gravity property on the android native control, and if the control has been disposed before unsubscribe the OnPropertyChangeHandler it crash. I think the PR #6467 will fix that.

**Comment by MitchBomcanhao** [comment]
I'm not sure if dispose is related to my particular variation of the issue.

To make it a bit clearer - I've got a custom grid control with an items source.
That items source, which had zero items before, is updated and will now try to render one item. This will fail when attempting to render the label in said template.

Relaunching the app and getting to the same page will render the control item without issues.


Edit:
I have however just tried replacing my custom grid control with a flexlayout (the items source and the template are exactly the same) and I'm not getting this crash anymore.
I'll do a bit more digging to see if I can find something in the original control that can be contributing to this.

**Comment by kvpt** [comment]
I finally succeed to reproduce the crash.
The issue was the dispose of the platform renderer which was deferred.
In function of timing, it can collide with the new page beeing layout.
So it can work fine, crash or make a blank screen.
I can only reproduce it with a splashscreen, I think it add a delay to the dispose process which make the issue more easily trigger.

The nuget is in the zip => [xamarin.forms.3.6.7.zip](https://github.com/xamarin/Xamarin.Forms/files/3279052/xamarin.forms.3.6.7.zip)

The PR #6467 will fix that case too.

**Comment by lassana** [comment]
Hmm, I still see the empty page with 3.6.7 however only when debugger is attached. If I start the app directly from the emulator, everything works fine.
[MyXamarinFormsApp-X.F-3.6.7.zip](https://github.com/xamarin/Xamarin.Forms/files/3279813/MyXamarinFormsApp-X.F-3.6.7.zip)

**Comment by kvpt** [comment]
Ok thanks,
So the good news is that I am on the good way.
But the bad news is that Android can call the destroy of the old activity in parallal with the creation of the new activity.
So I have to found a way to assure that destroy is executed before create in all case.

**Comment by kvpt** [comment]
So I made a new version, which use a synchronization primitive to prevent dispose and create to be executed in parallel. In theory it should fix the issue.
But as I can't reproduce the crash since my previous fix, I'm not able to confirm it fix the issue.
@lassana If you can try it [Xamarin.Forms.3.6.8.zip](https://github.com/xamarin/Xamarin.Forms/files/3283094/Xamarin.Forms.3.6.8.zip)

**Comment by lassana** [comment]
@kvpt still the same white screen.
Frankly, I don't think it is a concurrency issue, because the white screen appears every time regardless of how long I wait after the first page is closed. So, if I run the app, press the physical Back button, _wait for 30 seconds_, start the app again: anyway the white screen appears.

Like I said earlier, currently I can reproduce it only if debugger is attached. Weird.

FYI: I probably found a workaround (using X.F 3.6.8), at least for the project attached to the original post.

My first though was to add `NoHistory = true` to the Forms activity, and in fact it does fix the white screen and no crash occurs, however for some reason the Forms activity recreates twice. It happens with and without debugger. It looks like a X.F bug but different from what I reported.

<details>
 <summary>Screencast</summary>

![NoHistoryTrue mov](https://user-images.githubusercontent.com/2004960/59406573-566c9900-8db7-11e9-990c-2cf471cab67c.gif)

</details>

My next try was to use `LaunchMode = LaunchMode.SingleTask` or `LaunchMode = LaunchMode.SingleInstance` with the Forms activity, and, believe it or not, it fixes the issue.

<details>
 <summary>Screencast</summary>

![LaunchModeSingleTask mov](https://user-images.githubusercontent.com/2004960/59406843-207be480-8db8-11e9-9dad-536fe6b6217b.gif)

</details>

Is there any downsides for using `ActivityInfo.LAUNCH_SINGLE_TASK` `ActivityInfo.LAUNCH_SINGLE_INSTANCE` with Xamarin.Forms? I found [this post](https://xamarin.github.io/bugzilla-archives/42/42354/bug.html) by Jon Douglas but it was 3 years ago.

**Comment by kvpt** [comment]
@lassana This issue is caused by invalid dispose of the old renderers, that is the same as the new when the app reload when the application is static (this seem to be the case with mvvmcross).
I solve an issue of that type in #4707 (which is also include in this custom package).
So there is another cause that can trigger issue fixed by #4707.
The renderer dispose can also be caused by race condition, it was the case for me, so I deduced it was the same for you.

Anyway, thanks for your time and help me to fix this issue, it is very hard to reproduce it.

When you have the double call, can you please dump the call stacktrace of the OnCreate method.

For the launchmode I always use SingleTop which from documention is very similar to other SingleTask or SingleInstance if the Xamarin.Forms activity is at the top. For me it has work fine for years.
That said, I don't tested SingleTask or SingleInstance mode, the only issue that I can think about is if you have to use an other activity like a camera, I don't know how will react Xamarin.Forms in this case.
If you dont use any other activity in your app, I think it's safe.

**Comment by lassana** [comment]
If I use `SingleTop` or `NoHistory=True`, the Forms activity is shown twice. Here is the stacktrace of the `OnCreate` method:

<details>
 <summary>First launch (from VS for Mac)</summary>

```
MyXamarinFormsApp.Droid.MainActivity.OnCreate(Android.OS.Bundle bundle = (null)) in /Users/user/Desktop/MyXamarinFormsApp/src/MyXamarinFormsApp.Droid/Views/MainActivity.cs:27
Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(System.IntPtr jnienv = 0xffffffffb4059a80, System.IntPtr native__this = 0xffffffffbfbdf65c, System.IntPtr native_savedInstanceState = 0x0) in 
Android.Runtime.DynamicMethodNameCounter.3() in 
```

</details>

<details>
 <summary>Second launch (after the Back button pressed and the app icon tapped</summary>

```
MyXamarinFormsApp.Droid.MainActivity.OnCreate(Android.OS.Bundle bundle = (null)) in /Users/user/Desktop/MyXamarinFormsApp/src/MyXamarinFormsApp.Droid/Views/MainActivity.cs:27
Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(System.IntPtr jnienv = 0xffffffffb4059a80, System.IntPtr native__this = 0xffffffffbfbdf65c, System.IntPtr native_savedInstanceState = 0x0) in 
Android.Runtime.DynamicMethodNameCounter.3() in 
```

</details>

After the app icon is pressed in the system launcher `OnCreate` method is called only once (these Stacktraces are identical), however I see two Foms activities after second launch. Does it mean the first activity wasn't properly destroyed? That's really weird, because `OnPause`, `OnStop`, `OnDestroy` methods are called right after the physical Back button is pressed.

[MyXamarinFormsApp-XF-3.6.8.zip](https://github.com/xamarin/Xamarin.Forms/files/3285861/MyXamarinFormsApp-XF-3.6.8.zip)

**Comment by kvpt** [comment]
Ok I understand your new issue now.
It's not that you have two activities (the previous is destroy), it's you have now two pages.
I reproduce it with your sample but not I my app (which not use MvvmCross).
So I think that the issue is now either again in Xamarin.Forms or has moved to MvvmCross.
I need to check this.

**Comment by kvpt** [comment]
Ok, so after setting a test project I can confirm that the problem seem to be in MvvmCross.

On the first call, the page passed to the framework is a navigation page that contains only one HomePage.
![FirstCall](https://user-images.githubusercontent.com/1446221/59451509-e63c3280-8e0b-11e9-9113-9e8c21e7a5ee.png)

On the second, the page passed is also a navigation page but with two HomePage.
![SecondCall](https://user-images.githubusercontent.com/1446221/59451511-e76d5f80-8e0b-11e9-9a77-80f083b7b21a.png)

I will try to determine what cause this in MvvmCross but I promise nothing.

**Comment by kvpt** [comment]
The root cause is because the MvvmCross Framework in [MvxFormsPagePresenter.PushOrReplacePage](https://github.com/MvvmCross/MvvmCross/blob/c561b78fc2e1cd3e1f7f4699b705108da68f2c65/MvvmCross.Forms/Presenters/MvxFormsPagePresenter.cs#L644) recreate a page all the time and add it to the navigation page if it is the current root page.
I found that the process begin in [MvxAppStart.NavigateToFirstViewModel](https://github.com/MvvmCross/MvvmCross/blob/c561b78fc2e1cd3e1f7f4699b705108da68f2c65/MvvmCross/ViewModels/MvxAppStart.cs#L75) which call `NavigationService.Navigate`
The good news is that this framework is greatly architectured and provide a very good extensibility.
So this process can be replaced, the documentation is [here](https://www.mvvmcross.com/documentation/advanced/customizing-appstart). 

So to prevent the pushing of a new page at each start/restart, with overriding it's simple.

```csharp
public class MvxApp : MvxApplication
{
	public override void Initialize()
	{
		CreatableTypes()
			.EndingWith("Service")
			.AsInterfaces()
			.RegisterAsLazySingleton();

		//RegisterAppStart<HomeViewModel>(); <= To be replaced

		RegisterCustomAppStart<AppStart<HomeViewModel>>(); // <= With this
	}

	public class AppStart<TViewModel> : MvxAppStart<TViewModel> where TViewModel : IMvxViewModel
	{
		bool Loaded { get; set; }

		public AppStart(IMvxApplication application, IMvxNavigationService navigationService) 
			: base(application, navigationService)
		{
		}

		protected override async Task NavigateToFirstViewModel(object hint = null)
		{
			if (Loaded)
			{
				return;
			}

			await base.NavigateToFirstViewModel(hint);

			Loaded = true;
		}
	}
}
```

It's strange to me that the default behavior is like this, but it's perhaps like this because Xamarin.Forms was buggy.
If you want go further, you can perhaps use the MvvmCross [support](https://www.mvvmcross.com/support/).

**Comment by lassana** [comment]
Thank you a lot!

**Comment by jfversluis** [comment]
@kvpt @lassana is there still something that needs to be fixed because of this or is this something with MvvmCross? Can we close this issue?

**Comment by kvpt** [comment]
@jfversluis The crash initialy reported is solved by a fix that I have done in my fork.
I will make a PR during the week with explication of the root cause and the fix.
However, once this crash was solved MvvmCross need a tweak for handle the activity restart and don't recreate a new page.

**Comment by lassana** [comment]
@kvpt @jfversluis 
Since there is no crash with _the project from the OP + the NuGet from the post above + the MvvmCross workaround_ I think you can close it.
I tried X.F 4.0 and this workaround in one of our Xamarin app and I am still able to get a crash after the main page is changed two or three times. However, because X.F 4.0 doesn't include all your changes and this crash happens only when the debugger is attached, I don't think I can provide another minimal reproducible sample.
Waiting for a new stable X.F release with all these changes included!

**Comment by samhouts** [comment]
closed by #6596

**Comment by virsanna** [comment]
@jfversluis @samhouts @kvpt 
The same problem has appeared with Entry in case of Entry is within the TitleView. 
The app is crashed with "System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.EntryRenderer from native handle"
The slightly simplified code fragment is below. 
```
<NavigationPage.TitleView>
        <Grid ColumnDefinitions="*, Auto">
            <Entry x:Name="searchEntry"   
                   Grid.Column="0" 
                   Placeholder="Search..." Text="{Binding SearchTerm}"
                   TextChanged="searchEntry_TextChanged"
                   ReturnType="Search">
            </Entry>
            <Image Grid.Column="1" Source="close_icon.png">
            </Image>
        </Grid>
    </NavigationPage.TitleView>
```
Literally the same structure with Entry inside the Grid on the page itself does not cause a crash. 

Microsoft Visual Studio Community 2019
Version 16.8.4
VisualStudio.16.Release/16.8.4+30907.101
Microsoft .NET Framework
Version 4.8.04084

Xamarin   16.8.000.261 (d16-8@bb99248)
Xamarin Designer   16.8.0.507 (remotes/origin/d16-8@e87b24884)
Xamarin Templates   16.8.112 (86385a3)
Xamarin.Android SDK   11.1.0.26 (d16-8/a36ce73)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 5e9cb6d
    Java.Interop: xamarin/java.interop/d16-8@79d9533
    ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
    SQLite: xamarin/sqlite/3.32.1@1a3276b
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-8@2fb1cbc
Xamarin.iOS and Xamarin.Mac SDK   14.8.0.3 (c51fabee8)

**Comment by kvpt** [comment]
@virsanna This kind of issue is related to a failed dispose.
It can be caused by multiple causes and is very difficult and time consuming to debug and reproduce.
Are you able to to provide a stacktrace of the crash ?
Also I think it's better to open a new issue here, the crash is the same but the root cause (and so the fix) is different.

**Comment by virsanna** [comment]
@kvpt Thanks for the answer. Yes, I've realized from the discussion that it's caused by dispose. I didn't make the new issue, because the theme of this seems to me as completely appropriate for my case.
Here is the stacktrace from the AppCenter (this one is from the real device, and it was builded with an older version of XF, though not older than 4.7.*. But the VS Android emulator, with the parameters I've mentioned above, crashes absolutely the same way):

Xamarin Exception Stack:
System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.Platform.Android.EntryRenderer from native handle 0x7fc2f1e344 (key_handle 0x21165fa).
  at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x0017e] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at Java.Lang.Object.GetObject (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type type) [0x00023] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at Java.Lang.Object._GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00017] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at Java.Lang.Object.GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00000] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at Java.Lang.Object.GetObject[T] (System.IntPtr jnienv, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00006] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at Android.Text.ITextWatcherInvoker.n_AfterTextChanged_Landroid_text_Editable_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_s) [0x00000] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.100(intptr,intptr,intptr)
--- End of inner exception stack trace ---
  System.MissingMethodException: No constructor found for Xamarin.Forms.Platform.Android.EntryRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership)
    at Java.Interop.TypeManager.CreateProxy (System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x000b5] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
    at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00111] in <0ba9dbd9290b49e8bd855fff67eb1dc5>:0
  --- End of inner exception stack trace ---
    Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
      at Java.Lang.Error: Exception of type 'Java.Lang.Error' was thrown.
      at java.lang.Error: Java callstack:
      at crc643f46942d9dd1fff9.EntryRendererBase_1.n_afterTextChanged(Native Method)
      at crc643f46942d9dd1fff9.EntryRendererBase_1.afterTextChanged(Unknown Source:0)
      at android.widget.TextView.sendAfterTextChanged(TextView.java:11662)
      at android.widget.TextView$ChangeWatcher.afterTextChanged(TextView.java:14944)
      at android.text.SpannableStringBuilder.sendAfterTextChanged(SpannableStringBuilder.java:1287)
      at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:587)
      at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:517)
      at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:38)
      at android.view.inputmethod.BaseInputConnection.replaceText(BaseInputConnection.java:849)
      at android.view.inputmethod.BaseInputConnection.setComposingText(BaseInputConnection.java:618)
      at com.android.internal.view.IInputConnectionWrapper.executeMessage(IInputConnectionWrapper.java:393)
      at com.android.internal.view.IInputConnectionWrapper$MyHandler.handleMessage(IInputConnectionWrapper.java:89)
      at android.os.Handler.dispatchMessage(Handler.java:107)
      at android.os.Looper.loop(Looper.java:237)
      at android.app.ActivityThread.main(ActivityThread.java:8167)
      at java.lang.reflect.Method.invoke(Native Method)
      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)

Thread 2:
0   dalvik.system.VMStack.getThreadStackTrace(VMStack.java:-2)
1   java.lang.Thread.getStackTrace(Thread.java:1720)
2   java.lang.Thread.getAllStackTraces(Thread.java:1796)
3   com.microsoft.appcenter.crashes.Crashes.saveUncaughtException
4   com.microsoft.appcenter.crashes.WrapperSdkExceptionManager.saveWrapperException
5   crc643f46942d9dd1fff9.EntryRendererBase_1.n_afterTextChanged
6   crc643f46942d9dd1fff9.EntryRendererBase_1.afterTextChanged
7   android.widget.TextView.sendAfterTextChanged(TextView.java:11662)
8   android.widget.TextView$ChangeWatcher.afterTextChanged(TextView.java:14944)
9   android.text.SpannableStringBuilder.sendAfterTextChanged(SpannableStringBuilder.java:1287)
10  android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:587)
11  android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:517)
12  android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:38)
13  android.view.inputmethod.BaseInputConnection.replaceText(BaseInputConnection.java:849)
14  android.view.inputmethod.BaseInputConnection.setComposingText(BaseInputConnection.java:618)
15  com.android.internal.view.IInputConnectionWrapper.executeMessage(IInputConnectionWrapper.java:393)
16  com.android.internal.view.IInputConnectionWrapper$MyHandler.handleMessage(IInputConnectionWrapper.java:89)
17  android.os.Handler.dispatchMessage(Handler.java:107)
18  android.os.Looper.loop(Looper.java:237)
19  android.app.ActivityThread.main(ActivityThread.java:8167)
20  java.lang.reflect.Method.invoke(Method.java:-2)
21  com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
22  com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)

Thread 142311:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Object.wait(Object.java:442)
2   java.lang.Object.wait(Object.java:568)
3   java.lang.Daemons$ReferenceQueueDaemon.runInternal(Daemons.java:215)
4   java.lang.Daemons$Daemon.run(Daemons.java:137)
5   java.lang.Thread.run(Thread.java:919)

Thread 142312:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Object.wait(Object.java:442)
2   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:190)
3   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:211)
4   java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:271)
5   java.lang.Daemons$Daemon.run(Daemons.java:137)
6   java.lang.Thread.run(Thread.java:919)

Thread 142313:
0   java.lang.Thread.sleep(Thread.java:-2)
1   java.lang.Thread.sleep(Thread.java:440)
2   java.lang.Thread.sleep(Thread.java:356)
3   java.lang.Daemons$FinalizerWatchdogDaemon.sleepForMillis(Daemons.java:383)
4   java.lang.Daemons$FinalizerWatchdogDaemon.waitForFinalization(Daemons.java:411)
5   java.lang.Daemons$FinalizerWatchdogDaemon.runInternal(Daemons.java:323)
6   java.lang.Daemons$Daemon.run(Daemons.java:137)
7   java.lang.Thread.run(Thread.java:919)

Thread 142325:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:336)
2   android.os.Looper.loop(Looper.java:197)
3   android.os.HandlerThread.run(HandlerThread.java:67)

Thread 142327:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:336)
2   android.os.Looper.loop(Looper.java:197)
3   android.os.HandlerThread.run(HandlerThread.java:67)

Thread 142329:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:336)
2   android.os.Looper.loop(Looper.java:197)
3   android.os.HandlerThread.run(HandlerThread.java:67)

Thread 142334:
0   sun.misc.Unsafe.park(Unsafe.java:-2)
1   java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:230)
2   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2109)
3   java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:467)
4   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1091)
5   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
6   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
7   java.lang.Thread.run(Thread.java:919)

Thread 142338:
0   sun.misc.Unsafe.park(Unsafe.java:-2)
1   java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:230)
2   java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:461)
3   java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
4   java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:937)
5   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1091)
6   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
7   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
8   java.lang.Thread.run(Thread.java:919)

Thread 142339:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.ConnectionPool$1.run(ConnectionPool.java:106)
2   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
3   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
4   java.lang.Thread.run(Thread.java:919)

Thread 142340:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.okio.AsyncTimeout.awaitTimeout(AsyncTimeout.java:325)
2   com.android.okhttp.okio.AsyncTimeout.access$000(AsyncTimeout.java:42)
3   com.android.okhttp.okio.AsyncTimeout$Watchdog.run(AsyncTimeout.java:288)

Thread 142347:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:336)
2   android.os.Looper.loop(Looper.java:197)
3   android.os.HandlerThread.run(HandlerThread.java:67)

Thread 142348:
0   sun.misc.Unsafe.park(Unsafe.java:-2)
1   java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:230)
2   java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:461)
3   java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
4   java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:937)
5   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1091)
6   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
7   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
8   java.lang.Thread.run(Thread.java:919)

**Comment by kvpt** [comment]
@virsanna Thanks for the stacktrace, this is very helpful.
From the trace I think the problem is this line:
> at crc643f46942d9dd1fff9.EntryRendererBase_1.afterTextChanged(Unknown Source:0)

This indicate that the EntryRendererBase text change listener are not detached.
So I checked the code [here](https://github.com/xamarin/Xamarin.Forms/blob/5.0.0/Xamarin.Forms.Platform.Android/Renderers/EntryRenderer.cs), and I think I found the issue.

The listener is registered here when the Element is created.
https://github.com/xamarin/Xamarin.Forms/blob/130499628b44caabf6b3d8fc5e02e87c07c0b651/Xamarin.Forms.Platform.Android/Renderers/EntryRenderer.cs#L162
But is never disposed.
I will make some more tests to see if it doesn't break something and made a PR for that.
I really think it the cause of your crash.

**Comment by virsanna** [comment]
@kvpt Yes, it seems like true. May be, you can also use the following. 
The page is rendered as it has to, even the breakpoint in TextChanged listener while the page initial rendering is reached, but then, when the execution is continued and I put the cursor to the Entry wanting to enter the text, the app crashes.

**Comment by kvpt** [comment]
@virsanna While waiting for the next service release you can test the latest [nightly build](https://github.com/xamarin/Xamarin.Forms/wiki/Nightly-Builds) which include the fix that I made.
I think it will fix your issue.

**Comment by kvpt** [comment]
@virsanna The fix is now include in the service release 3 of the 5.0 branch.

**Comment by virsanna** [comment]
Thanks a lot, @kvpt !
